### PR TITLE
DistributeRewardsCallback Test

### PR DESCRIPTION
**Pull request by phamminh0811**
## 1. Summary
Fixes #617 

## 2.Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## 5. Checklist

- [ ] Does the Readme need to be updated?

### CODE REVIEW

File: x/interchainstaking/keeper/callbacks_test.go [hunk 1]
```diff
@@ -762,6 +766,228 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+
+			if test.pass {
+				suite.NoError(err)
+			} else {
+				suite.Error(err)
+			}
+
+			test.check()
+			zone, _ = quicksilver.InterchainstakingKeeper.GetZone(ctxA, suite.chainB.ChainID)
+			fmt.Println("redemption rate: ", zone.RedemptionRate)
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/callbacks_test.go:977]: are these outputs required, or just debugging

File: x/interchainstaking/keeper/callbacks_test.go [hunk 2]
```diff
@@ -762,6 +766,228 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+
+			commitments := quicksilver.IBCKeeper.ChannelKeeper.GetAllPacketCommitments(ctxA)
+			fmt.Println("commitments: ", commitments[0])
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/callbacks_test.go:980]: are these outputs required, or just debugging

File: x/interchainstaking/keeper/callbacks_test.go [hunk 3]
```diff
@@ -762,6 +766,228 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+				return respbz
+			},
+			queryMsg: icqtypes.Query{ChainId: suite.chainB.ChainID},
+			check:    func() {},
+			pass:     false,
+		},
+	}
+	for _, test := range tests {
+		suite.Run(test.name, func() {
+			fmt.Println("redemption rate: ", zone.RedemptionRate)
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/callbacks_test.go:947]: are these outputs required, or just debugging

File: x/interchainstaking/keeper/callbacks_test.go [hunk 4]
```diff
@@ -762,6 +766,228 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+			}
+
+			test.check()
+			zone, _ = quicksilver.InterchainstakingKeeper.GetZone(ctxA, suite.chainB.ChainID)
+			fmt.Println("redemption rate: ", zone.RedemptionRate)
+
+			commitments := quicksilver.IBCKeeper.ChannelKeeper.GetAllPacketCommitments(ctxA)
+			fmt.Println("commitments: ", commitments[0])
+
+			channel, found := quicksilver.IBCKeeper.ChannelKeeper.GetChannel(ctxA, icstypes.TransferPort, channelID)
```
Comment (joe-bowman) [inline x/interchainstaking/keeper/callbacks_test.go:982]: What is the purpose of closing the channel here?

**Comment by joe-bowman** [comment]
A couple of questions, and also linting required.

**Comment by phamminh0811** [comment]
Oh, my bad. I forgot deleting debug stuffs. Will fix it later

File: x/interchainstaking/keeper/callbacks_test.go (hunk 4 above)
Comment (phamminh0811) [reply x/interchainstaking/keeper/callbacks_test.go:972]: there is a test case named "no connection setup" that require there is no transfer channel existed in order to make it work

**Comment by joe-bowman** [comment]
Just linting ser :)

**Comment by phamminh0811** [comment]
Done, please check

File: x/interchainstaking/keeper/callbacks_test.go [hunk 5]
```diff
@@ -762,6 +767,219 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+	suite.setupTestZones()
+	quicksilver := suite.GetQuicksilverApp(suite.chainA)
+	gaia := suite.GetQuicksilverApp(suite.chainB)
+	quicksilver.InterchainstakingKeeper.CallbackHandler().RegisterCallbacks()
+
+	ctxA := suite.chainA.GetContext()
+	ctxB := suite.chainB.GetContext()
+	vals := gaia.StakingKeeper.GetAllValidators(ctxB)
+
+	zone, _ := quicksilver.InterchainstakingKeeper.GetZone(ctxA, suite.chainB.ChainID)
```
Comment (ThanhNhann) [inline x/interchainstaking/keeper/callbacks_test.go:781]: Should check require `found` here

File: x/interchainstaking/keeper/callbacks_test.go [hunk 6]
```diff
@@ -762,6 +767,219 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+		name            string
+		zoneSetup       func()
+		connectionSetup func() string
+		responseMsg     func() []byte
+		queryMsg        icqtypes.Query
+		check           func()
+		pass            bool
+	}{
+		{
+			name: "valid case with positive rewards and -5% < delta < 2%",
```
Comment (ThanhNhann) [inline x/interchainstaking/keeper/callbacks_test.go:798]: I think we should have some details about the calculation of this testcase

File: x/interchainstaking/keeper/callbacks_test.go [hunk 7]
```diff
@@ -762,6 +767,219 @@ func (suite *KeeperTestSuite) TestHandleValideRewardsCallback() {
...
+			check: func() {
+				zone, _ := quicksilver.InterchainstakingKeeper.GetZone(ctxA, suite.chainB.ChainID)
+				redemptionRate := zone.RedemptionRate
+
+				suite.Equal(prevRedemptionRate, redemptionRate)
+			},
+			pass: true,
+		},
+		{
+			name:      "invalid host zone",
```
Comment (ThanhNhann) [inline x/interchainstaking/keeper/callbacks_test.go:871]: I think we should update the name of test case to `Invalid chainId`

File: x/interchainstaking/keeper/callbacks_test.go (hunk 4 above)
Comment (phamminh0811) [reply x/interchainstaking/keeper/callbacks_test.go:985]: right now I set the context of the chain in the beginning of the test, so each test case will reuse the same context and reuse the channel that already been created before. that could make test case wrong so we should close channel after finish each test case

**Comment by ThanhNhann** [comment]
Plz fix the conflict then we can merge this pr. We also have to make interchaintest for this function @anhductn2001

**Review by ThanhNhann — APPROVED** [review]
Lgtm!

**Review by joe-bowman — APPROVED** [review]
lgtm

**Pull request merged**